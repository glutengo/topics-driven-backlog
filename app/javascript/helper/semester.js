import moment from 'moment';

const ssMonth = 4;
const wsMonth = 10;
const startDay = 1;

export const WS = 'W';
export const SS = 'S';

/**
 * Find when the start of a semester was from a given semester, month and year
 * @param semester
 * @param month
 * @param year
 * @returns {*}
 */
function findSemesterStart(semester, month, year) {
    // if we're already in the new year, semester started last year
    if (semester === WS && month < ssMonth) {
        year -= 1;
    }

    return moment().set({
        year,
        month: (semester === WS ? wsMonth : ssMonth) - 1,
        date: startDay,
    });
}

/**
 * A human-readable string for the current year(s) of the semester
 * @param semester
 * @param year
 * @returns {string}
 */
function getYearString(semester, year) {
    return year.toString().substring(2)
        + (semester === WS ? '/' + (year + 1).toString().substring(2) : '');
}

/**
 * A human-readable string for the current semester
 * @param semester
 * @returns {string}
 */
function getSemesterString(semester) {
    return semester === WS ? 'WS' : 'SS';
}

/**
 * A human-readable string for the current semester + year(s)
 * @param semester
 * @param year
 * @returns {string}
 */
function getFullString(semester, year) {
    return getSemesterString(semester) + ' ' + getYearString(semester, year);
}

/**
 * An object with the calculated semester data from a given moment
 * @param date
 * @returns {{semesterStart: *, semester: string, year: number, fullString: string, valueString: string}}
 */
function calculateSemester(date) {
    const year = date.year();
    const month = date.month() + 1; // 0-indexed

    const semester = month >= ssMonth && month < wsMonth ? SS : WS;

    const semesterStart = findSemesterStart(semester, month, year);
    const semesterStartYear = semesterStart.year();

    return {
        semesterStart,
        semester,
        year: semesterStartYear,
        fullString: getFullString(semester, semesterStartYear),
        valueString: semester + '*' + semesterStartYear
    };
}

/**
 * Calculate the current semester
 * @returns {{semesterStart: *, semester: string, year: number, fullString: string, valueString: string}}
 */
export function current() {
    return calculateSemester(moment());
}

/**
 * Calculate the next semester
 * @returns {{semesterStart: *, semester: string, year: number, fullString: string, valueString: string}}
 */
export function next() {
    return calculateSemester(moment().add(6, 'months'));
}

export default {WS, SS, current, next};