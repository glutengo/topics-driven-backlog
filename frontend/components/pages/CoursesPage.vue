<template>
    <section id="courses-page">
        <h1>Courses</h1>

        <ul 
            v-if="courses.length" 
            class="list-unstyled">

            <CourseItem
                class="mb-2"
                v-for="course in courses"
                :data="course"
                :key="course.id"/>
        </ul>

        <p v-else>No courses yet.</p>

        <CourseForm
            v-if="showForm"
            v-model="newCourse"
            @cancel="showForm = false"
            @submit="saveCourse"/>

        <BButton
            v-else
            type="button"
            variant="primary"
            @click="showForm = true">Add course</BButton>
    </section>
</template>

<script>
import BForm from '@bootstrap/form/form';
import BFormGroup from '@bootstrap/form-group/form-group';
import BFormInput from '@bootstrap/form-input/form-input';
import BFormSelect from '@bootstrap/form-select/form-select';
import BRow from '@bootstrap/layout/row';
import BCol from '@bootstrap/layout/col';
import BButton from '@bootstrap/button/button';
import CourseForm from '@/components/forms/CourseForm';
import CourseItem from '@/components/elements/CourseItem';

export default {
    components: { BForm, BFormGroup, BFormInput, BFormSelect, BRow, BCol, BButton, CourseForm, CourseItem },
    data() {
        return {
            showForm: false,
            newCourse: this.$store.getters['courses/template'](),
        };
    },
    computed: {
        /**
         * All courses
         * @returns {array}
         */
        courses() {
            return this.$store.getters['courses/all'];
        },
    },
    methods: {
        /**
         * Save current form state as new story
         */
        async saveCourse() {
            const semesterSplit = this.newCourse.semester.split('*');
            this.newCourse.semester_type = semesterSplit[0];
            this.newCourse.semester_year = parseInt(semesterSplit[1], 10);

            await this.$store.dispatch('courses/create', this.newCourse);

            this.newCourse = this.$store.getters['courses/template']();

            // TODO handle errors in UI
        },
    },
};
</script>