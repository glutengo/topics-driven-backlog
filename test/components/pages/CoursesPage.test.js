import { mount } from 'vue-test-utils';
import CoursesPage from '@/components/pages/CoursesPage';
import { current } from '@/helper/semester';

describe('CoursesPage.test.js', () => {
    let cmp, dispatch = jest.fn();

    beforeEach(() => {
        cmp = mount(CoursesPage, {
            mocks: {
                $store: {
                    dispatch,
                    state: {
                        user: {
                            permissions: {
                                courses: {
                                    create: true,
                                    read: true,
                                },
                            },
                        },
                    },
                    getters: {
                        'courses/all': () => [
                            {
                                title: 'Test course',
                                semester_type: 'W',
                                semester_year: 2017,
                                permissions: {

                                }
                            },
                        ],
                        'courses/template': () => {
                            return {
                                semester: current().valueString,
                                allow_enrollment: true,
                            };
                        },
                    },
                },
            },
        });

        jest.resetAllMocks();
    });

    it('adds a course', () => {
        cmp.vm.newCourse = {
            title: 'Test course',
            semester: 'W*2017',
        };

        cmp.vm.saveCourse();

        expect(dispatch).toBeCalled();
    });
});
