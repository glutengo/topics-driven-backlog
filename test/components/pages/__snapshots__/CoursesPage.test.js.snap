// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoursesPage.test.js has the expected html structure 1`] = `
<section
  id="courses-page"
>
  <h1>
    Courses
  </h1>
   
  <ul
    class="list-unstyled"
  >
    <li
      class="card mb-3"
    >
      <div
        class="card-body"
      >
        <router-link
          class="h4 card-title link-unstyled"
          to="/courses/undefined"
        >
          
                    
                
        </router-link>
         
        <h5
          class="h6 text-muted card-subtitle mt-0"
        >
          
                    Test course (WS 17/18)
                
        </h5>
         
        <!---->
      </div>
    </li>
  </ul>
   
  <button
    class="btn btn-primary"
    type="button"
  >
    Add course
  </button>
</section>
`;

exports[`CoursesPage.test.js has the expected html structure when form visible 1`] = `
<section
  id="courses-page"
>
  <h1>
    Courses
  </h1>
   
  <ul
    class="list-unstyled"
  >
    <li
      class="card mb-3"
    >
      <div
        class="card-body"
      >
        <router-link
          class="h4 card-title link-unstyled"
          to="/courses/undefined"
        >
          
                    
                
        </router-link>
         
        <h5
          class="h6 text-muted card-subtitle mt-0"
        >
          
                    Test course (WS 17/18)
                
        </h5>
         
        <!---->
      </div>
    </li>
  </ul>
   
  <form
    class=""
  >
    <div
      class="row"
    >
      <div
        class="col-md-4"
      >
        <div
          aria-labelledby="__BVID__36___BV_label_"
          class="b-form-group form-group"
          id="__BVID__36_"
          role="group"
        >
          <label
            class="col-form-label pt-0"
            for="course-shorthand"
            id="__BVID__36___BV_label_"
          >
            Shorthand
          </label>
          <div
            class=""
          >
            <input
              aria-required="true"
              class="form-control"
              id="course-shorthand"
              required="required"
              type="text"
            />
            <!---->
            <!---->
            <!---->
          </div>
        </div>
      </div>
       
      <div
        class="col-md-8"
      >
        <div
          aria-labelledby="__BVID__38___BV_label_"
          class="b-form-group form-group"
          id="__BVID__38_"
          role="group"
        >
          <label
            class="col-form-label pt-0"
            for="course-title"
            id="__BVID__38___BV_label_"
          >
            Title
          </label>
          <div
            class=""
          >
            <input
              aria-required="true"
              class="form-control"
              id="course-title"
              required="required"
              type="text"
            />
            <!---->
            <!---->
            <!---->
          </div>
        </div>
      </div>
    </div>
     
    <div
      aria-labelledby="__BVID__40___BV_label_"
      class="b-form-group form-group"
      id="__BVID__40_"
      role="group"
    >
      <label
        class="col-form-label pt-0"
        for="course-hyperlink"
        id="__BVID__40___BV_label_"
      >
        Link
      </label>
      <div
        class=""
      >
        <input
          class="form-control"
          id="course-hyperlink"
          type="text"
        />
        <!---->
        <!---->
        <!---->
      </div>
    </div>
     
    <div
      aria-labelledby="__BVID__42___BV_label_"
      class="b-form-group form-group"
      id="__BVID__42_"
      role="group"
    >
      <label
        class="col-form-label pt-0"
        for="course-semester"
        id="__BVID__42___BV_label_"
      >
        Semester
      </label>
      <div
        class=""
      >
        <select
          class="form-control custom-select"
          id="course-semester"
        >
          <option
            value="W*2017"
          >
            WS 17/18
          </option>
           
          <option
            value="S*2018"
          >
            SS 18
          </option>
        </select>
        <!---->
        <!---->
        <!---->
      </div>
    </div>
     
    <button
      class="btn btn-primary disabled"
      disabled="disabled"
      type="submit"
    >
      Save
    </button>
     
    <button
      class="btn btn-secondary"
      type="button"
    >
      Cancel
    </button>
  </form>
</section>
`;
